smart-contracts/wasm-transform/src/artifact.rs

```diff
@@ -670,124 +717,517 @@ impl BackPatchStack {
         let lookup_idx = self.stack.len() - n as usize - 1;
         self.stack.get_mut(lookup_idx).ok_or_else(|| anyhow!("Attempt to access unknown label."))
     }
+
+    pub fn get(&self, n: LabelIndex) -> CompileResult<&JumpTarget> {
+        ensure!(
+            (n as usize) < self.stack.len(),
+            "Attempt to access label beyond the size of the stack."
+        );
+        let lookup_idx = self.stack.len() - n as usize - 1;
+        self.stack.get(lookup_idx).ok_or_else(|| anyhow!("Attempt to access unknown label."))
+    }
+}
+
+/// A generator of dynamic locations needed in addition to the locals during
+/// execution.
+struct DynamicLocations {
+    /// The next location to give out if there are no reusable locations.
+    next_location:      i32,
+    /// A set of locations that are available for use again.
+    reusable_locations: BTreeSet<i32>,
```

Comment:
It is not important to have the smallest one, no. But it is important that we don't have duplicates in this collection and with the stack I think we'd have to be manually checking that in some cases when we call "reuse" since the same location can appear multiple times on the operand stack.

Reply:
I'll add a comment.

Reply:
We also have to remove from the collection sometimes, not necessarily from the top.